File: ws/src/sample_vision/scripts/image_publisher.py
#!/usr/bin/env python
# coding:utf-8

import rospy

from sensor_msgs.msg import Image
import cv2
from cv_bridge import CvBridge

cvBridge = CvBridge()


def imageCallback(msg):
    if not isinstance(msg, Image):
        return

    # cv bridge
    # Ros Image -> OpenCv Mat
    mat = cvBridge.imgmsg_to_cv2(msg, "bgr8")
    cv2.imshow("pymat", mat)
    cv2.waitKey(3)

    # OpenCv Mat -> Ros Image
    imgmsg = cvBridge.cv2_to_imgmsg(mat, "bgr8")
    publisher.publish(imgmsg)


if __name__ == '__main__':
    # 创建节点
    nodeName = "image_subscriber"
    rospy.init_node(nodeName)

    pubName = "/haha"
    publisher = rospy.Publisher(pubName, Image, queue_size=1000)

    topicName = "/usb_cam/image_raw"
    rospy.Subscriber(topicName, Image, imageCallback)

    # 阻塞线程
    rospy.spin()

File: ws/src/sample_vision/src/image_publisher.cpp
#include <iostream>
#include "ros/ros.h"

#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include "sensor_msgs/Image.h"

using namespace std;

ros::Publisher *publisher = nullptr;
uint seq = 0;

void imageCallback(const sensor_msgs::Image::ConstPtr &msg) {
  //ros 图像数据格式
//  sensor_msgs::Image --> cv::Mat

  const cv_bridge::CvImageConstPtr &ptr = cv_bridge::toCvShare(msg, "bgr8");
  cv::Mat mat = ptr->image;

  cv::imshow("mat", mat);
  cv::waitKey(3);

  // cv::Mat --> sensor_msgs::Image -> 传输出去
  cv_bridge::CvImage image;
//  image.data = cv_bridge::
  image.header.stamp = ros::Time::now();
  image.header.seq = ++seq;
  image.encoding = "bgr8";
  image.image = mat;
  publisher->publish(image.toImageMsg());
}

int main(int argc, char **argv) {
  //初始化节点
  string nodeName = "image_subscriber";
  ros::init(argc, argv, nodeName);
  //创建节点
  ros::NodeHandle node;

  string pubName = "/hello";
  ros::Publisher pub = node.advertise<sensor_msgs::Image>(pubName, 1000);
  publisher = &pub;

  string topicName = "/usb_cam/image_raw";
  ros::Subscriber subscriber = node.subscribe(topicName, 1000, imageCallback);


  //阻塞
  ros::spin();
  return 0;
}